daml-script/daml3/Daml/Script/Questions/Submit.daml

```diff
@@ -10,168 +11,258 @@ module Daml.Script.Questions.Submit
+data Submission = Submission with
+  sActAs : NonEmpty Party
+  sReadAs : [Party]
+  sDisclosures : [Disclosure]
+  sErrorBehaviour : ErrorBehaviour
+  sCommands : [CommandWithMeta]
+  sLocation : Optional (Text, SrcLoc)
+
+data ErrorBehaviour = MustSucceed | MustFail | Try
+
+-- | Options to detemine the stakeholders of a transaction, as well as disclosures.
+-- Intended to be specified using the `actAs`, `readAs` and `disclose` builders, combined using the Semigroup concat `(<>)` operator.
+-- ```
+-- actAs alice <> readAs [alice, bob] <> disclose myContract
+-- ```
+-- Note that actAs and readAs follows the same party derivation rules as `signatory`, see their docs for examples.
+-- All submissions must specify at least one `actAs` party, else a runtime error will be thrown.
+-- A minimum submission may look like
+-- ```
+-- submit (actAs alice) $ createCmd MyContract with party = alice
+-- ```
+data SubmitOptions = SubmitOptions with
+  soActAs : [Party]
+  soReadAs : [Party]
+  soDisclosures : [Disclosure]
+
+-- | Semigroup instance allowing for combination via (<>)
+instance Semigroup SubmitOptions where
+  SubmitOptions a b c <> SubmitOptions a' b' c' = SubmitOptions (a <> a') (b <> b') (c <> c')
+
+-- | Builds a SubmitOptions with given actAs parties.
+-- Any given submission must include at least one actAs party.
+-- Note that the parties type is constrainted by `IsParties`, allowing for specifying parties as any of the following:
+-- ```
+-- Party
+-- [Party]
+-- NonEmpty Party
+-- Set Party
+-- Optional Party
+-- ```
+actAs : IsParties p => p -> SubmitOptions
+actAs p = SubmitOptions (toParties p) [] []
+
+-- | Builds a SubmitOptions with given readAs parties.
+-- A given submission may omit any readAs parties and still be valid.
+-- Note that the parties type is constrainted by `IsParties`, allowing for specifying parties as any of the following:
+-- ```
+-- Party
+-- [Party]
+-- NonEmpty Party
+-- Set Party
+-- Optional Party
+-- ```
+readAs : IsParties p => p -> SubmitOptions
+readAs p = SubmitOptions [] (toParties p) []
+
+-- | Provides many Explicit Disclosures to the transaction.
+discloseMany : [Disclosure] -> SubmitOptions
+discloseMany ds = SubmitOptions [] [] ds
+
+-- | Provides an Explicit Disclosure to the transaction.
+disclose : Disclosure -> SubmitOptions
+disclose d = discloseMany [d]
+
+-- | Defines a type that can be transformed into a SubmitOptions
+class IsSubmitOptions a where
+  toSubmitOptions : a -> SubmitOptions
+
+-- Even though `IsParties SubmitOptions` doesn't exist, ghc doesn't consider context in instance selection.
+instance {-# OVERLAPS #-} IsSubmitOptions SubmitOptions where
+  toSubmitOptions = identity
+
+-- | Allows a single (or many) parties to be directly provided as actAs parties.
+-- This gives rise to the familiar syntax of
+-- ```
+-- alice `submit` createCmd MyContract with party = alice
+-- ```
+instance IsParties p => IsSubmitOptions p where
+  toSubmitOptions = actAs
```

Comment:
This is unlikely but if someone wanted to add their own instance elsewhere they would need to mark it as `{-# OVERLAP{S,PING} #-}`, otherwise the compiler wouldn't know what to pick between that new instance and `instance IsParties p => IsSubmitOptions p`. Instead, I would make the more `IsParties p =>` instance `{-# OVERLAPPABLE #-}` and leave `instance IsSubmitOptions SubmitOptions` without a pragma.

```suggestion
instance IsSubmitOptions SubmitOptions where
  toSubmitOptions = identity

-- | Allows a single (or many) parties to be directly provided as actAs parties.
-- This gives rise to the familiar syntax of
-- ```
-- alice `submit` createCmd MyContract with party = alice
-- ```
-- This is OVERLAPPABLE to allow GHC to prefer more specific instances if
-- available (e.g. instance 'IsSubmitOptions SubmitOptions' above), since GHC
-- doesn't consider context in instance selection.
instance {-# OVERLAPPABLE #-} IsParties p => IsSubmitOptions p where
  toSubmitOptions = actAs
```

https://downloads.haskell.org/ghc/latest/docs/users_guide/exts/instances.html#overlapping-instances

Reply:
Ah right yeah, thats the correct way to do this, good point

Reply:
Opted for explicit instances, as per your other comment